tests/unit/s2n_crypto_test.c

```diff
@@ -0,0 +1,174 @@
+                    S2N_ERR_NULL);
+        };
+
+        /* Test: successfully get master secret */
+        {
+            uint8_t output[S2N_TLS_SECRET_LEN] = { 0 };
+            EXPECT_SUCCESS(s2n_connection_get_master_secret(conn, output, sizeof(output)));
+            EXPECT_BYTEARRAY_EQUAL(test_secret, output, sizeof(output));
+        };
+
+        /* Test: TLS1.3 not supported */
+        {
+            uint8_t output[S2N_TLS_SECRET_LEN] = { 0 };
+
+            conn->actual_protocol_version = S2N_TLS13;
+            EXPECT_FAILURE_WITH_ERRNO(
+                    s2n_connection_get_master_secret(conn, output, sizeof(output)),
+                    S2N_ERR_INVALID_STATE);
+
+            conn->actual_protocol_version = S2N_TLS12;
+            EXPECT_SUCCESS(s2n_connection_get_master_secret(conn, output, sizeof(output)));
+            EXPECT_BYTEARRAY_EQUAL(test_secret, output, sizeof(output));
+        };
+
+        /* Test: at least S2N_TLS_SECRET_LEN of output required */
+        {
+            uint8_t output[S2N_TLS_SECRET_LEN] = { 0 };
+
+            /* Fail if insufficient memory */
+            EXPECT_FAILURE_WITH_ERRNO(
+                    s2n_connection_get_master_secret(conn, output, 0),
+                    S2N_ERR_INSUFFICIENT_MEM_SIZE);
+            EXPECT_FAILURE_WITH_ERRNO(
+                    s2n_connection_get_master_secret(conn, output, 1),
+                    S2N_ERR_INSUFFICIENT_MEM_SIZE);
+            EXPECT_FAILURE_WITH_ERRNO(
+                    s2n_connection_get_master_secret(conn, output, S2N_TLS_SECRET_LEN - 1),
+                    S2N_ERR_INSUFFICIENT_MEM_SIZE);
+
+            /* Succeed if exactly S2N_TLS_SECRET_LEN bytes */
+            EXPECT_SUCCESS(s2n_connection_get_master_secret(conn, output, S2N_TLS_SECRET_LEN));
+            EXPECT_BYTEARRAY_EQUAL(test_secret, output, sizeof(output));
+
+            /* Succeed if more than S2N_TLS_SECRET_LEN bytes */
+            EXPECT_SUCCESS(s2n_connection_get_master_secret(conn, output, S2N_TLS_SECRET_LEN + 1));
+            EXPECT_BYTEARRAY_EQUAL(test_secret, output, sizeof(output));
+        };
+
+        /* Test: handshake must be complete */
+        {
+            uint8_t output[S2N_TLS_SECRET_LEN] = { 0 };
+
+            conn->handshake.message_number = 0;
+            EXPECT_FAILURE_WITH_ERRNO(
+                    s2n_connection_get_master_secret(conn, output, sizeof(output)),
+                    S2N_ERR_HANDSHAKE_NOT_COMPLETE);
+
+            EXPECT_OK(s2n_skip_handshake(conn));
+            EXPECT_SUCCESS(s2n_connection_get_master_secret(conn, output, sizeof(output)));
+            EXPECT_BYTEARRAY_EQUAL(test_secret, output, sizeof(output));
+        };
+
+        /* Test: self-talk */
+        for (size_t i = 0; i < s2n_array_len(supported_versions); i++) {
+            const uint8_t version = supported_versions[i];
+            if (s2n_is_in_fips_mode() && version == S2N_SSLv3) {
```

Comment:
nit: It might be nice to add a comment on why this restriction is necessary.

Reply:
It's unrelated to what we're testing here and probably a bug: https://github.com/aws/s2n-tls/issues/4476 At least, the first blocker I found was clearly a bug.